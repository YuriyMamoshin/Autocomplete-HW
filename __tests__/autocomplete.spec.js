import React from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";

import Autocomplete from "../src/app/components/autocomplete";
global.fetch = jest.fn(() => Promise.resolve({ data: [] }));

const renderElementWithClient = (element, client) => {
  return render(
    <QueryClientProvider client={client}>{element}</QueryClientProvider>
  );
};

describe("<Autocomplete /", () => {
  const queryClient = new QueryClient();

  beforeEach(() => {
    fetch.mockClear();
  });

  it("should render Autocomplete component", () => {
    const { asFragment } = renderElementWithClient(
      <Autocomplete />,
      queryClient
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call API when there more than three characters provided", async () => {
    renderElementWithClient(<Autocomplete />, queryClient);
    const input = screen.getByPlaceholderText("Type your request");

    fireEvent.change(input, { target: { value: "cats and dogs" } });

    await waitFor(() => {
      expect(fetch).toHaveBeenCalled();
    });
  });

  it("should not call API when there less than three characters provided", async () => {
    renderElementWithClient(<Autocomplete />, queryClient);
    const input = screen.getByPlaceholderText("Type your request");

    fireEvent.change(input, { target: { value: "ca" } });

    await waitFor(() => {
      expect(fetch).not.toHaveBeenCalled();
    });
  });
});
