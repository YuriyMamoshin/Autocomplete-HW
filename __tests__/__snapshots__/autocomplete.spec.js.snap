// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete / should render Autocomplete component 1`] = `
<DocumentFragment>
  <section
    class="autocomplete"
  >
    <input
      class="autocomplete__input"
      placeholder="Type your request"
      type="text"
      value=""
    />
  </section>
</DocumentFragment>
`;
